README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

## d3rain

[![Travis build
status](https://travis-ci.org/daranzolin/d3rain.svg?branch=master)](https://travis-ci.org/daranzolin/d3rain)

According to the authorities at Urban Dictionary, [‘drip’ is synonymous
with ‘immense
swag.’](https://www.urbandictionary.com/define.php?term=Drip) This
package brings some D3 drip to R.

### Installation

You can install `d3rain` from GitHub via:

``` r
remotes::install_github("daranzolin/d3rain")
```

## Examples

‘Rain’ visualizations are useful aids to observe the relationship
between a ranked, numeric variable (e.g. percentile, rank, etc.) and any
factored, categorical variable.

``` r
library(d3rain)
library(dplyr)

armed_levels <- c('No', 'Knife', 'Non-lethal firearm', 'Firearm')
pk <- fivethirtyeight::police_killings %>% 
  filter(armed %in% armed_levels,
         !is.na(age)) %>% 
  mutate(armed = factor(armed, levels = armed_levels)) 

pk %>% 
  d3rain(age, armed, toolTip = raceethnicity, title = "2015 Police Killings by Age, Armed Status") %>% 
  drip_settings(dripSequence = 'iterate',
                ease = 'bounce',
                jitterWidth = 20,
                dripSpeed = 1000,
                dripFill = 'firebrick') %>% 
  chart_settings(fontFamily = 'times',
                 yAxisTickLocation = 'left')
```

![Alt
Text](https://raw.githubusercontent.com/daranzolin/d3rain/master/inst/img/d3raingif1.gif)

`drip_settings` adjusts the drip sequence, easing animation, jitter
width, drip speed, and color. `chart_settings` controls the axes
displays, font size, font family, and background color.

You can adjust the drip iteration by reordering the data frame:

``` r
pk %>% 
  arrange(age) %>% 
  d3rain(age, armed, toolTip = raceethnicity, title = "2015 Police Killings by Age, Armed Status") %>% 
  drip_settings(dripSequence = 'iterate',
                ease = 'linear',
                jitterWidth = 25,
                dripSpeed = 500,
                dripFill = 'steelblue') %>% 
  chart_settings(fontFamily = 'times',
                 yAxisTickLocation = 'left')
```

![Alt
Text](https://raw.githubusercontent.com/daranzolin/d3rain/master/inst/img/d3raingif2.gif)

## Future Work

  - Additional drip behaviors (e.g. by group)
  - Conditional fill colors


NEWS.md
# d3rain 0.0.1
* Submitting to CRAN (2019-05-11)


R/d3rain.R
#' Create a d3rain visualization
#'
#' @param .data A table of data
#' @param x A numeric 'ranking' variable, e.g. percentile, rank, etc.
#' @param y A factored, ordinal variable
#' @param toolTip Which variable to display on drip tooltips
#' @param reverseX Whether to reverse the x-axis
#' @param title Visualization title
#'
#' @import htmlwidgets
#'
#' @export
#' @examples
#' mtcars$cyl <- factor(mtcars$cyl)
#' mtcars$car <- rownames(mtcars)
#' d3rain(mtcars, mpg, cyl, car)
d3rain <- function(.data, x, y, toolTip, reverseX = FALSE, title = '') {

  x <- rlang::enquo(x)
  y <- rlang::enquo(y)
  toolTip <- rlang::enquo(toolTip)

  cx <- all(.data[,rlang::quo_name(toolTip)] == .data[,rlang::quo_name(x)])
  cy <- all(.data[,rlang::quo_name(toolTip)] == .data[,rlang::quo_name(y)])

  out_df <- as.data.frame(subset(.data, select = c(tidyselect::vars_select(names(.data), !!x, !!y, !!toolTip))))

  if (cx) {
    out_df$toolTip <- out_df[,1]
    toolTipName <- names(out_df)[1]
  } else if (cy) {
    out_df$toolTip <- out_df[,2]
    toolTipName <- names(out_df)[2]
  } else {
    toolTipName <- names(out_df)[3]
  }

  names(out_df) <- c('ind', 'group', 'toolTip')
  if (!is.numeric(out_df$ind)) stop ("x must be numeric.", call. = FALSE)
  if (!is.factor(out_df$group)) stop("y must be a factor.", call. = FALSE)

  x = list(
    data = out_df,
    toolTipName = toolTipName,
    y_domain = levels(out_df$group),
    title = title,
    reverseX = reverseX
  )

  # create widget
  htmlwidgets::createWidget(
    name = 'd3rain',
    x,
    package = 'd3rain'
  )
}

#' Adjust drip style and behavior
#'
#' @param d3rain An object of class d3rain
#' @param dripFill Color of drips
#' @param dripOpacity Opacity of drips
#' @param dripSequence Either 'iterate' or 'together'
#' @param ease Either 'bounce' or 'linear'
#' @param dripSpeed Drip speed in milliseconds
#' @param iterationSpeedX Iteration speed multiplier
#' @param jitterWidth Jitter width in pixels along x-axis
#'
#' @export
#' @examples
#' mtcars$cyl <- factor(mtcars$cyl)
#' mtcars$car <- rownames(mtcars)
#' d3rain(mtcars, mpg, cyl, car) %>%
#'     drip_settings(ease = 'linear', jitterWidth = 25, dripSpeed = 500)
drip_settings <- function(d3rain,
                          dripFill = 'firebrick',
                          dripOpacity = 0.5,
                          dripSequence = 'iterate',
                          ease = 'bounce',
                          dripSpeed = 1500,
                          iterationSpeedX = 100,
                          jitterWidth = 0) {

  if (!any(dripSequence %in% c('iterate', 'together', 'by_group'))) {
    stop("dripSequence param must be 'iterate' or 'together'", call. = FALSE)
  }
  if (!inherits(d3rain, 'd3rain')) {
    stop("d3rain must be of class 'd3rain'")
  }
  if (!any(ease %in% c('bounce', 'linear'))) {
    stop("ease param must be 'bounce' or 'linear'", call. = FALSE)
  }
  if (!any(c(is.numeric(dripSpeed), is.numeric(iterationSpeedX)))) {
    stop("dripSpeed and iterationSpeedX must be numeric", call. = FALSE)
  }

  d3rain$x$dripFill <- dripFill
  d3rain$x$dripOpacity <- dripOpacity
  d3rain$x$dripSequence <- dripSequence
  d3rain$x$ease <- ease
  d3rain$x$dripSpeed <- dripSpeed
  d3rain$x$iterationSpeedX <- iterationSpeedX
  d3rain$x$jitterWidth <- jitterWidth
  return(d3rain)
}

#' Adjust chart settings
#'
#' @param d3rain An object of class d3rain
#' @param toolTipTextColor Color of tooltip text
#' @param backgroundFill Background color of SVG
#' @param fontSize Font size
#' @param fontFamily Font family, e.g. 'times', 'sans-serif'
#' @param yAxisTickLocation Location of y-axis ticks, either 'center', 'left', or 'right'
#'
#' @return d3rain
#' @export
#' @examples
#' mtcars$cyl <- factor(mtcars$cyl)
#' mtcars$car <- rownames(mtcars)
#' d3rain(mtcars, mpg, cyl, car) %>%
#'     drip_settings(ease = 'linear', jitterWidth = 25, dripSpeed = 500) %>%
#'     chart_settings(fontFamily = 'times', yAxisTickLocation = 'left')
chart_settings <- function(d3rain,
                       toolTipTextColor = 'white',
                       backgroundFill = 'white',
                       fontSize = 18,
                       fontFamily = 'sans-serif',
                       yAxisTickLocation = 'center') {

  if (!yAxisTickLocation %in% c('left', 'right', 'center')) {
    stop("yAxisTickLocation param must be either 'left', 'right', or 'center'", call. = FALSE)
  }

  d3rain$x$backgroundFill <- backgroundFill
  d3rain$x$fontSize <- fontSize
  d3rain$x$fontFamily <- fontFamily
  d3rain$x$yAxisTickLocation <- yAxisTickLocation
  d3rain$x$toolTipTextColor <- toolTipTextColor
  return(d3rain)
}


#' Shiny bindings for d3rain
#'
#' Output and render functions for using d3rain within Shiny
#' applications and interactive Rmd documents.
#'
#' @param outputId output variable to read from
#' @param width,height Must be a valid CSS unit (like \code{'100\%'},
#'   \code{'400px'}, \code{'auto'}) or a number, which will be coerced to a
#'   string and have \code{'px'} appended.
#' @param expr An expression that generates a d3rain
#' @param env The environment in which to evaluate \code{expr}.
#' @param quoted Is \code{expr} a quoted expression (with \code{quote()})? This
#'   is useful if you want to save an expression in a variable.
#'
#' @name d3rain-shiny
#'
#' @export
d3rainOutput <- function(outputId, width = '100%', height = '400px'){
  htmlwidgets::shinyWidgetOutput(outputId, 'd3rain', width, height, package = 'd3rain')
}

#' @rdname d3rain-shiny
#' @export
renderD3rain <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) { expr <- substitute(expr) } # force quoted
  htmlwidgets::shinyRenderWidget(expr, d3rainOutput, env, quoted = TRUE)
}


inst/htmlwidgets/lib/d3-tip/index-min.js
(function(root,factory){if(typeof define==="function"&&define.amd){define(["d3"],factory)}else if(typeof module==="object"&&module.exports){var d3=require("d3");module.exports=factory(d3)}else{root.d3.tip=factory(root.d3)}})(this,function(d3){return function(){var direction=d3_tip_direction,offset=d3_tip_offset,html=d3_tip_html,node=initNode(),svg=null,point=null,target=null;function tip(vis){svg=getSVGNode(vis);point=svg.createSVGPoint();document.body.appendChild(node)}tip.show=function(){var args=Array.prototype.slice.call(arguments);if(args[args.length-1]instanceof SVGElement)target=args.pop();var content=html.apply(this,args),poffset=offset.apply(this,args),dir=direction.apply(this,args),nodel=getNodeEl(),i=directions.length,coords,scrollTop=document.documentElement.scrollTop||document.body.scrollTop,scrollLeft=document.documentElement.scrollLeft||document.body.scrollLeft;nodel.html(content).style("opacity",1).style("pointer-events","all");while(i--)nodel.classed(directions[i],false);coords=direction_callbacks.get(dir).apply(this);nodel.classed(dir,true).style("top",coords.top+poffset[0]+scrollTop+"px").style("left",coords.left+poffset[1]+scrollLeft+"px");return tip};tip.hide=function(){var nodel=getNodeEl();nodel.style("opacity",0).style("pointer-events","none");return tip};tip.attr=function(n,v){if(arguments.length<2&&typeof n==="string"){return getNodeEl().attr(n)}else{var args=Array.prototype.slice.call(arguments);d3.selection.prototype.attr.apply(getNodeEl(),args)}return tip};tip.style=function(n,v){if(arguments.length<2&&typeof n==="string"){return getNodeEl().style(n)}else{var args=Array.prototype.slice.call(arguments);d3.selection.prototype.style.apply(getNodeEl(),args)}return tip};tip.direction=function(v){if(!arguments.length)return direction;direction=v==null?v:functor(v);return tip};tip.offset=function(v){if(!arguments.length)return offset;offset=v==null?v:functor(v);return tip};tip.html=function(v){if(!arguments.length)return html;html=v==null?v:functor(v);return tip};tip.destroy=function(){if(node){getNodeEl().remove();node=null}return tip};function d3_tip_direction(){return"n"}function d3_tip_offset(){return[0,0]}function d3_tip_html(){return" "}var direction_callbacks=d3.map({n:direction_n,s:direction_s,e:direction_e,w:direction_w,nw:direction_nw,ne:direction_ne,sw:direction_sw,se:direction_se}),directions=direction_callbacks.keys();function direction_n(){var bbox=getScreenBBox();return{top:bbox.n.y-node.offsetHeight,left:bbox.n.x-node.offsetWidth/2}}function direction_s(){var bbox=getScreenBBox();return{top:bbox.s.y,left:bbox.s.x-node.offsetWidth/2}}function direction_e(){var bbox=getScreenBBox();return{top:bbox.e.y-node.offsetHeight/2,left:bbox.e.x}}function direction_w(){var bbox=getScreenBBox();return{top:bbox.w.y-node.offsetHeight/2,left:bbox.w.x-node.offsetWidth}}function direction_nw(){var bbox=getScreenBBox();return{top:bbox.nw.y-node.offsetHeight,left:bbox.nw.x-node.offsetWidth}}function direction_ne(){var bbox=getScreenBBox();return{top:bbox.ne.y-node.offsetHeight,left:bbox.ne.x}}function direction_sw(){var bbox=getScreenBBox();return{top:bbox.sw.y,left:bbox.sw.x-node.offsetWidth}}function direction_se(){var bbox=getScreenBBox();return{top:bbox.se.y,left:bbox.e.x}}function initNode(){var node=d3.select(document.createElement("div"));node.style("position","absolute").style("top",0).style("opacity",0).style("pointer-events","none").style("box-sizing","border-box");return node.node()}function getSVGNode(el){el=el.node();if(el.tagName.toLowerCase()==="svg")return el;return el.ownerSVGElement}function getNodeEl(){if(node===null){node=initNode();document.body.appendChild(node)}return d3.select(node)}function getScreenBBox(){var targetel=target||d3.event.target;while("undefined"===typeof targetel.getScreenCTM&&"undefined"===targetel.parentNode){targetel=targetel.parentNode}var bbox={},matrix=targetel.getScreenCTM(),tbbox=targetel.getBBox(),width=tbbox.width,height=tbbox.height,x=tbbox.x,y=tbbox.y;point.x=x;point.y=y;bbox.nw=point.matrixTransform(matrix);point.x+=width;bbox.ne=point.matrixTransform(matrix);point.y+=height;bbox.se=point.matrixTransform(matrix);point.x-=width;bbox.sw=point.matrixTransform(matrix);point.y-=height/2;bbox.w=point.matrixTransform(matrix);point.x+=width;bbox.e=point.matrixTransform(matrix);point.x-=width/2;point.y-=height/2;bbox.n=point.matrixTransform(matrix);point.y+=height;bbox.s=point.matrixTransform(matrix);return bbox}function functor(v){return typeof v==="function"?v:function(){return v}}return tip}});


inst/htmlwidgets/d3rain.js
HTMLWidgets.widget({

  name: 'd3rain',

  type: 'output',

  factory: function(el, width, height) {

    return {

      renderValue: function(opts) {

        margin = ({top: 100, right: 25, bottom: 20, left: 25});

        const data = HTMLWidgets.dataframeToD3(opts.data);
        //console.log(data);

        const svg = d3.select(el)
                    .append("svg")
                    .style("width", "100%")
                    .style("height", "100%");

        svg.append("rect")
            .attr("width", "100%")
            .attr("height", "100%")
            .attr("fill", opts.hasOwnProperty('backgroundFill') ? opts.backgroundFill : 'white');

        let yAxisTickLocation;
        let textAnchor;
        switch(opts.yAxisTickLocation) {
          case 'center':
            yAxisTickLocation = width/2;
            textAnchor = 'middle';
          break;
          case 'right':
            yAxisTickLocation = width - margin.right;
            textAnchor = 'end';
          break;
          case 'left':
            yAxisTickLocation = margin.right;
            textAnchor = 'start';
          break;
          default:
        }

        let fontSize = opts.hasOwnProperty('fontSize') ? opts.fontSize : 18;
        let fontFamily = opts.hasOwnProperty('fontFamily') ? opts.fontFamily : 'sans-serif';
        let jitterWidth = opts.hasOwnProperty('jitterWidth') ? opts.jitterWidth : 0;
        let dripSequence = opts.hasOwnProperty('dripSequence') ? opts.dripSequence : 'iterate';

        function ease(o) {
            return o === 'bounce' ? d3.easeBounce : d3.easeLinear;
        }

        let x;
        if (opts.reverseX) {
          x = d3.scaleLinear()
                .domain([d3.max(data, d => +d.ind), d3.min(data, d => +d.ind)])
                .range([margin.left, width - margin.right]);
        } else {
          x = d3.scaleLinear()
                .domain([d3.min(data, d => +d.ind), d3.max(data, d => +d.ind)])
                .range([margin.left, width - margin.right]);
        }

        let xAxis = g => g
                .style("font", `${opts.fontSize}px ${opts.fontFamily}`)
                .call(d3.axisTop(x)
                    .tickPadding(15)
                    .tickValues(d3.extent(data, d => +d.ind))
                    .tickSize(0))
                .call(g => g.select(".domain").remove());

        let y = d3.scaleBand()
                .domain(opts.y_domain)
                .range([margin.top, height - margin.bottom]);

        let yAxis = g => g
                .attr("transform", `translate(${margin.left},0)`)
                .call(d3.axisLeft(y))
                .call(g => g.selectAll(".tick line").clone()
                  .attr("stroke-opacity", 0.2)
                  .attr("x2", width - margin.right))
                  .attr('transform', 'translate(0, 10)')
                .call(g => g.select(".domain").remove());

        svg.append("g")
              .attr('transform', `translate(0, ${margin.top/2 + 5})`)
              .call(xAxis);

        svg.append("g")
                .call(yAxis)
                .selectAll("text")
                .attr("x", yAxisTickLocation)
                .attr("text-anchor", textAnchor)
                .attr('transform', 'translate(0, 10)')
                .style("font", `${opts.fontSize}px ${opts.fontFamily}`);

        svg.append("text")
                .attr("x", (width / 2))
                .attr("y", margin.top / 6)
                .attr("text-anchor", "middle")
                .style("font-family", `${fontFamily}`)
                .style("font-size", "20px")
                .style("font-weight", "bold")
                .text(opts.title);


        let tip = d3.tip()
              .attr('class', 'd3-tip')
              .offset([7, 7])
              .html(function(d) {
                return `<span style = "color:${opts.toolTipTextColor}">${opts.toolTipName}: ${d.toolTip}</span>`;
              });

        svg.call(tip);

        let circles = svg.selectAll('circle')
            .data(data)
              .enter().append('circle')
                .attr('cx', d => x(d.ind))
                .attr('cy', margin.top / 2)
                .attr('r', 5)
                .attr('fill', opts.hasOwnProperty('dripFill') ? opts.dripFill : 'firebrick')
                .style('opacity', opts.hasOwnProperty('dripOpacity') ? opts.dripOpacity : 0.5)
                .on('mouseover', tip.show)
                .on('mouseout', tip.hide);


        if (dripSequence === 'iterate') {
          d3.timeout(_ => {
            circles = circles.transition()
            .delay((d, i) => opts.hasOwnProperty('iterationSpeedX') ? opts.iterationSpeedX * i : 100 * i)
            .duration(opts.hasOwnProperty('dripSpeed') ? opts.dripSpeed : 1000)
            .ease(opts.hasOwnProperty('ease') ? ease(opts.ease) : d3.easeBounce)
            .attr('cy', d => y(d.group) + y.bandwidth() / 2 - Math.random() * jitterWidth);
            }, 1500);
        } else {
          d3.timeout(_ => {
            circles = circles.transition()
            .duration(opts.hasOwnProperty('dripSpeed') ? opts.dripSpeed : 1000)
            .ease(opts.hasOwnProperty('ease') ? ease(opts.ease) : d3.easeBounce)
            .attr('cy', d => y(d.group) + y.bandwidth() / 2 - Math.random() * jitterWidth);
          }, 1500);
        }
      },

      resize: function(width, height) {

        // TODO: code to re-render the widget with a new size

      }

    };
  }
});
